source/extensions/http/header_validators/envoy_default/path_normalizer.cc

```diff
@@ -0,0 +1,268 @@
+  // delimiters. The only exception is for percent-encoded octets corresponding
+  // to characters in the unreserved set, which can be decoded at any time.
+  // SPELLCHECKER(on)
+  //
+  char ch;
+
+  if (!isxdigit(octet[1]) || !isxdigit(octet[2])) {
+    return {PercentDecodeResult::Invalid};
+  }
+
+  // normalize to UPPERCASE
+  octet[1] = octet[1] >= 'a' && octet[1] <= 'z' ? octet[1] ^ 0x20 : octet[1];
+  octet[2] = octet[2] >= 'a' && octet[2] <= 'z' ? octet[2] ^ 0x20 : octet[2];
+
+  // decode to character
+  ch = octet[1] >= 'A' ? (octet[1] - 'A' + 10) : (octet[1] - '0');
+  ch *= 16;
+  ch += octet[2] >= 'A' ? (octet[2] - 'A' + 10) : (octet[2] - '0');
+
+  if (testChar(kUnreservedCharTable, ch)) {
+    // Based on RFC, only decode characters in the UNRESERVED set.
+    return {PercentDecodeResult::Decoded, ch};
+  }
+
+  if (ch == '/' || ch == '\\') {
+    // We decoded a slash character and how we handle it depends on the active configuration.
+    switch (config_.uri_path_normalization_options().path_with_escaped_slashes_action()) {
+    case HeaderValidatorConfig_UriPathNormalizationOptions::IMPLEMENTATION_SPECIFIC_DEFAULT:
+      ABSL_FALLTHROUGH_INTENDED;
+    case HeaderValidatorConfig_UriPathNormalizationOptions::KEEP_UNCHANGED:
+      // default implementation: normalize the encoded octet and accept the path
+      return {PercentDecodeResult::Normalized};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::REJECT_REQUEST:
+      // Reject the entire request
+      return {PercentDecodeResult::Reject};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::UNESCAPE_AND_FORWARD:
+      // Decode the slash and accept the path.
+      return {PercentDecodeResult::Decoded, ch};
+
+    case HeaderValidatorConfig_UriPathNormalizationOptions::UNESCAPE_AND_REDIRECT:
+      // Decode the slash and response with a redirect to the normalized path.
+      return {PercentDecodeResult::DecodedRedirect, ch};
+
+    default:
+      // This should never occur but it's here to make the compiler happy because of the extra
+      // values added by protobuf.
+      break;
+    }
+  }
+
+  // The octet is a valid encoding but it wasn't be decoded because it was outside the UNRESERVED
+  // character set.
+  return {PercentDecodeResult::Normalized};
+}
+
+HeaderValidator::RequestHeaderMapValidationResult
+PathNormalizer::normalizePathUri(RequestHeaderMap& header_map) const {
+  // Make a copy of the original path so we can edit it in place.
+  absl::string_view original_path = header_map.path();
+  size_t length = original_path.size();
+  auto path_ptr = std::make_unique<char[]>(length + 1); // auto free on return
+  char* path = path_ptr.get();
+
+  original_path.copy(path, length);
+
+  // We rely on the string being null terminated so that we can safely look forward 1 character.
+  path[length] = '\0';
+
+  // Start normalizing the path.
+  char* read = path;
```

Comment:
It would be preferable to use iterators instead of pointers, since in debug builds you get bounds validation and in release builds it works just as a pointer. You can allocate std::string for destination instead of raw char array, so it you get destination iterator and absl::string_view has iterator support as well.

Reply:
Done. I switched to using iterators and `string_view` objects rather than raw pointers.